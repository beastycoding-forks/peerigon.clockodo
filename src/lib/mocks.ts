import { faker } from "@faker-js/faker";

export const ONE_DAY = 24 * 60 * 60 * 1000;
const MAX_ITERATION_COUNT = 10000;

export const isoDateFromDateTime = (dateTime: Date) => {
  return dateTime.toISOString().replace(/T.*/, "");
};

export const isoMonthFromDateTime = (dateTime: Date) => {
  return dateTime.toISOString().replace(/-\d\dT.*/, "");
};

export const isoDateFromTimestamp = (timestamp: number) => {
  return isoDateFromDateTime(new Date(timestamp));
};

export const timestampFromIsoDate = (isoDate: string) => {
  return new Date(isoDate).getTime();
};

export const endOfMonth = (date: Date) => {
  let startOfNextMonth = new Date(date.getTime());

  startOfNextMonth.setMonth(date.getMonth() + 1);

  if (startOfNextMonth.getMonth() === 0) {
    startOfNextMonth.setFullYear(date.getFullYear() + 1);
  }

  startOfNextMonth = new Date(isoMonthFromDateTime(startOfNextMonth));

  const endOfMonth = new Date(startOfNextMonth);

  endOfMonth.setTime(endOfMonth.getTime() - ONE_DAY / 2);

  return new Date(isoDateFromDateTime(endOfMonth));
};

export const endOfYear = (dateTime: Date) => {
  return new Date(`${dateTime.getFullYear()}-12-31T23:59:59.999Z`);
};

export const nextDay = (date: Date) => {
  return new Date(isoDateFromDateTime(new Date(date.getTime() + ONE_DAY)));
};

export const generateUnique = <Value>({
  count,
  generate,
}: {
  count: number;
  generate: () => Value;
}) => {
  const generated = new Set<Value>();

  for (let i = 0; generated.size < count; i++) {
    if (i === MAX_ITERATION_COUNT) {
      throw new Error(
        `Couldn't generate enough unique values before reaching the max iteration count.
This usually happens when the generated values create too many conflicts (e.g. too many dates on a small date range).`
      );
    }
    generated.add(generate());
  }

  return generated;
};

export const generateDayRange = ({
  count,
  between: [from, to],
}: {
  count: number;
  between: [Date, Date];
}) => {
  return Array.from(
    generateUnique({
      count,
      generate: () => {
        const randomDay = isoDateFromDateTime(faker.date.between(from, to));

        return timestampFromIsoDate(randomDay);
      },
    })
  ).sort();
};

export const generateMonthRange = ({
  count,
  between: [from, to],
}: {
  count: number;
  between: [Date, Date];
}) => {
  return Array.from(
    generateUnique({
      count,
      generate: () => {
        const randomMonth = isoMonthFromDateTime(faker.date.between(from, to));

        return timestampFromIsoDate(randomMonth);
      },
    })
  ).sort();
};

export const toPairs = <Item>(array: Array<Item>) => {
  if (array.length % 2 !== 0) {
    throw new Error("Cannot create pairs: Array length must be divisible by 2");
  }

  const count = array.length / 2;
  const pairs = new Array<[Item, Item]>(count);

  for (let i = 0; i < count; i++) {
    pairs[i] = [array[i * 2], array[i * 2 + 1]];
  }

  return pairs;
};
