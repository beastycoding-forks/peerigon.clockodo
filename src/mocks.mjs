// eslint-disable-next-line import/no-unresolved
export * from "./mocks.js";
