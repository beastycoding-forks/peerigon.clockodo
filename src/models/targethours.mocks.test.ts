import { createTargethoursRowMocks } from "./targethours.mocks.js";

describe("createTargethoursRowMocks()", () => {
  test("It creates realistic mocks", () => {
    expect(createTargethoursRowMocks({ count: 20 })).toMatchSnapshot();
  });
});
