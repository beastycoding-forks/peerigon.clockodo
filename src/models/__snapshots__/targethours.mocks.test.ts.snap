// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTargethoursRowMocks() It creates realistic mocks 1`] = `
Array [
  Object {
    "compensationMonthly": 7,
    "dateSince": "2019-03-01",
    "dateUntil": "2019-04-30",
    "id": 0,
    "monthlyTarget": 34317,
    "type": "monthly",
    "usersId": 0,
    "workdayFriday": false,
    "workdayMonday": true,
    "workdaySaturday": false,
    "workdaySunday": true,
    "workdayThursday": false,
    "workdayTuesday": true,
    "workdayWednesday": false,
  },
  Object {
    "compensationMonthly": 3,
    "dateSince": "2019-05-01",
    "dateUntil": "2019-06-30",
    "id": 1,
    "monthlyTarget": 71233,
    "type": "monthly",
    "usersId": 0,
    "workdayFriday": true,
    "workdayMonday": true,
    "workdaySaturday": false,
    "workdaySunday": true,
    "workdayThursday": false,
    "workdayTuesday": false,
    "workdayWednesday": true,
  },
  Object {
    "compensationMonthly": 6,
    "dateSince": "2019-07-01",
    "dateUntil": "2019-09-30",
    "id": 2,
    "monthlyTarget": 23451,
    "type": "monthly",
    "usersId": 0,
    "workdayFriday": false,
    "workdayMonday": false,
    "workdaySaturday": true,
    "workdaySunday": false,
    "workdayThursday": true,
    "workdayTuesday": true,
    "workdayWednesday": true,
  },
  Object {
    "absenceFixedCredit": true,
    "compensationDaily": 26,
    "compensationMonthly": 3,
    "dateSince": "2020-01-13",
    "dateUntil": "2020-02-18",
    "friday": 0.96,
    "id": 3,
    "monday": 24,
    "saturday": 3,
    "sunday": 24,
    "thursday": 7,
    "tuesday": 0,
    "type": "weekly",
    "usersId": 0,
    "wednesday": 0,
  },
  Object {
    "absenceFixedCredit": true,
    "compensationDaily": 27,
    "compensationMonthly": 0,
    "dateSince": "2020-02-19",
    "dateUntil": "2020-03-18",
    "friday": 5,
    "id": 4,
    "monday": 0,
    "saturday": 3.34,
    "sunday": 0,
    "thursday": 24,
    "tuesday": 0,
    "type": "weekly",
    "usersId": 0,
    "wednesday": 2,
  },
  Object {
    "absenceFixedCredit": false,
    "compensationDaily": 19,
    "compensationMonthly": 2,
    "dateSince": "2020-04-15",
    "dateUntil": "2020-04-30",
    "friday": 8,
    "id": 5,
    "monday": 24,
    "saturday": 0,
    "sunday": 24,
    "thursday": 3,
    "tuesday": 0,
    "type": "weekly",
    "usersId": 0,
    "wednesday": 5,
  },
  Object {
    "absenceFixedCredit": true,
    "compensationDaily": 9,
    "compensationMonthly": 8,
    "dateSince": "2020-06-13",
    "dateUntil": "2020-06-26",
    "friday": 24,
    "id": 6,
    "monday": 1.55,
    "saturday": 1,
    "sunday": 1.55,
    "thursday": 0,
    "tuesday": 0,
    "type": "weekly",
    "usersId": 0,
    "wednesday": 4,
  },
  Object {
    "absenceFixedCredit": false,
    "compensationDaily": 17,
    "compensationMonthly": 8,
    "dateSince": "2020-06-27",
    "dateUntil": "2020-08-10",
    "friday": 24,
    "id": 7,
    "monday": 0,
    "saturday": 3,
    "sunday": 0,
    "thursday": 0,
    "tuesday": 0.12,
    "type": "weekly",
    "usersId": 0,
    "wednesday": 8,
  },
  Object {
    "absenceFixedCredit": false,
    "compensationDaily": 22,
    "compensationMonthly": 6,
    "dateSince": "2020-08-11",
    "dateUntil": "2020-10-25",
    "friday": 24,
    "id": 8,
    "monday": 3,
    "saturday": 4,
    "sunday": 3,
    "thursday": 2.55,
    "tuesday": 0,
    "type": "weekly",
    "usersId": 0,
    "wednesday": 0,
  },
  Object {
    "absenceFixedCredit": false,
    "compensationDaily": 33,
    "compensationMonthly": 7,
    "dateSince": "2020-11-02",
    "dateUntil": "2020-11-03",
    "friday": 0,
    "id": 9,
    "monday": 0,
    "saturday": 4,
    "sunday": 0,
    "thursday": 2.86,
    "tuesday": 24,
    "type": "weekly",
    "usersId": 0,
    "wednesday": 7,
  },
  Object {
    "absenceFixedCredit": true,
    "compensationDaily": 53,
    "compensationMonthly": 5,
    "dateSince": "2020-11-06",
    "dateUntil": "2020-12-01",
    "friday": 24,
    "id": 10,
    "monday": 4,
    "saturday": 4,
    "sunday": 4,
    "thursday": 0,
    "tuesday": 2.85,
    "type": "weekly",
    "usersId": 0,
    "wednesday": 0,
  },
  Object {
    "absenceFixedCredit": true,
    "compensationDaily": 33,
    "compensationMonthly": 2,
    "dateSince": "2020-12-02",
    "dateUntil": "2020-12-22",
    "friday": 0,
    "id": 11,
    "monday": 24,
    "saturday": 2,
    "sunday": 24,
    "thursday": 7,
    "tuesday": 0,
    "type": "weekly",
    "usersId": 0,
    "wednesday": 1.92,
  },
  Object {
    "absenceFixedCredit": true,
    "compensationDaily": 49,
    "compensationMonthly": 2,
    "dateSince": "2020-12-23",
    "dateUntil": "2020-12-25",
    "friday": 8,
    "id": 12,
    "monday": 0,
    "saturday": 1,
    "sunday": 0,
    "thursday": 0,
    "tuesday": 2.57,
    "type": "weekly",
    "usersId": 0,
    "wednesday": 24,
  },
  Object {
    "absenceFixedCredit": true,
    "compensationDaily": 3,
    "compensationMonthly": 5,
    "dateSince": "2020-12-28",
    "dateUntil": "2021-01-05",
    "friday": 0,
    "id": 13,
    "monday": 24,
    "saturday": 2.53,
    "sunday": 24,
    "thursday": 7,
    "tuesday": 2,
    "type": "weekly",
    "usersId": 0,
    "wednesday": 0,
  },
  Object {
    "absenceFixedCredit": true,
    "compensationDaily": 39,
    "compensationMonthly": 4,
    "dateSince": "2021-01-06",
    "dateUntil": "2021-02-16",
    "friday": 3,
    "id": 14,
    "monday": 4.63,
    "saturday": 6,
    "sunday": 4.63,
    "thursday": 0,
    "tuesday": 24,
    "type": "weekly",
    "usersId": 0,
    "wednesday": 0,
  },
  Object {
    "absenceFixedCredit": true,
    "compensationDaily": 34,
    "compensationMonthly": 3,
    "dateSince": "2021-03-29",
    "dateUntil": "2021-05-11",
    "friday": 24,
    "id": 15,
    "monday": 6,
    "saturday": 1,
    "sunday": 6,
    "thursday": 0,
    "tuesday": 0,
    "type": "weekly",
    "usersId": 0,
    "wednesday": 2.34,
  },
  Object {
    "absenceFixedCredit": false,
    "compensationDaily": 10,
    "compensationMonthly": 6,
    "dateSince": "2021-05-12",
    "dateUntil": "2021-06-30",
    "friday": 0,
    "id": 16,
    "monday": 0,
    "saturday": 3,
    "sunday": 0,
    "thursday": 24,
    "tuesday": 6.46,
    "type": "weekly",
    "usersId": 0,
    "wednesday": 6,
  },
  Object {
    "absenceFixedCredit": false,
    "compensationDaily": 54,
    "compensationMonthly": 8,
    "dateSince": "2021-07-01",
    "dateUntil": "2021-10-11",
    "friday": 0,
    "id": 17,
    "monday": 4,
    "saturday": 7.87,
    "sunday": 4,
    "thursday": 24,
    "tuesday": 5,
    "type": "weekly",
    "usersId": 0,
    "wednesday": 0,
  },
  Object {
    "absenceFixedCredit": true,
    "compensationDaily": 3,
    "compensationMonthly": 8,
    "dateSince": "2021-10-12",
    "dateUntil": "2021-10-31",
    "friday": 0,
    "id": 18,
    "monday": 0.74,
    "saturday": 0,
    "sunday": 0.74,
    "thursday": 4,
    "tuesday": 24,
    "type": "weekly",
    "usersId": 0,
    "wednesday": 3,
  },
  Object {
    "absenceFixedCredit": false,
    "compensationDaily": 10,
    "compensationMonthly": 0,
    "dateSince": "2021-11-01",
    "dateUntil": null,
    "friday": 0,
    "id": 19,
    "monday": 24,
    "saturday": 6,
    "sunday": 24,
    "thursday": 0,
    "tuesday": 5.94,
    "type": "weekly",
    "usersId": 0,
    "wednesday": 4,
  },
]
`;
